File: Python-Src/node_tron.py
from node import Node
from layer import Layer
from network import Network
from state import State
from pipe import Pipe
from math import exp, pow
import random

class NodeTron(Network):

    def __init__(self):
        Network.__init__(self)

    def build_network(self):
        main_node = Node(2, 1, 'mainNode')
        hidden_layer = Layer(self.learn_rate)
        self.input_layer = [State(), State()]
        self.output_layer = [State()]

        for state in self.input_layer:
            self.connect_state_to_node(state, main_node)
        for state in self.output_layer:
            self.connect_node_to_state(main_node, state)
        
        hidden_layer.add_node(main_node)

        self.layers.append(hidden_layer)
        
        print("finish building")


    def train_network(self, iterations=0, cases=[]):
        while iterations > 0:
            inst = cases[random.randint(0, len(cases) - 1)]
            self.run_forward([inst[0], inst[1]])
            self.run_backward([inst[2]])
            iterations -= 1

File: Csharp-Src/Csharp-Src/NodeTron.cs
using System;
using System.Collections.Generic;

namespace Csharp_Src
{
    public class NodeTron : Network
    {
        public List<List<double>> AndCases { get; }
        public List<List<double>> OrCases { get; }
        public NodeTron() : base()
        {
            this.AndCases = new List<List<double>>()
            {
                new List<double>() {1.0, 1.0, 1.0},
                new List<double>() {1.0, 0.0, 0.0},
                new List<double>() {0.0, 1.0, 0.0},
                new List<double>() {0.0, 0.0, 0.0},
            };
            this.OrCases = new List<List<double>>()
            {
                new List<double>() {1.0, 1.0, 1.0},
                new List<double>() {0.0, 1.0, 1.0},
                new List<double>() {0.0, 0.0, 0.0},
                new List<double>() {1.0, 0.0, 1.0},
            };
        }

        public override void BuildNetwork()
        {
            Node mainNode = new Node(2, 1, "mainNode");
            Layer hiddenLayer = new Layer(this.LearningRate);
            List<State> inputLayer = new List<State>() { new State(), new State() };
            List<State> outputLayer = new List<State>() { new State() };

            foreach (var state in inputLayer)
                this.ConnectStateToNode(state, mainNode);
            foreach (var state in outputLayer)
                this.ConnectNodeToState(mainNode, state);

            hiddenLayer.AddNode(mainNode);

            this.Layers.Add(hiddenLayer);
            this.OutputLayer = outputLayer;
            this.InputLayer = inputLayer;
        }

        public override void TrainNetwork(int iterations, List<List<double>> cases)
        {
            Random rand = new Random();
            while (iterations > 0)
            {
                List<double> inst = cases[rand.Next(0, cases.Count)];
                this.RunForward(new List<double> { inst[0], inst[1] });
                this.RunBackward(new List<double> { inst[2] });
                iterations--;
            }
        }
    }
}
